File: COMP9041/lab06/all_whales.py
#!/usr/bin/env python

# reimplement all_wales using python
import sys, re
whale_record = sys.stdin.readlines();

# scan each line
whales_namelist = []
whales_odd = {}
whales_ind = {}
for pod in whale_record:

    # split the pod by number and strings
    pod = pod.strip() # remove the spaces
    splitted = pod.split(" ") # note that we can not use " +"
    splitted = filter(None, splitted) # remove those ""
    
    individual_ob = int(splitted[0])
    whale_name = ' '.join(splitted[1:])
    whale_name = whale_name.lower() # all lower case
    whale_name = whale_name.strip('\n') # remove \n
    if whale_name[-1] == 's':
        whale_name = whale_name[:-1] # remove s at last
    
    # check whether this whale has been observed
    if whale_name in whales_odd:
        whales_odd[whale_name] += 1
        whales_ind[whale_name] += individual_ob
    else: # new whale
        whales_namelist.append(whale_name)
        whales_odd[whale_name] = 1
        whales_ind[whale_name] = individual_ob

# output the data in alphabetical order
whales_namelist = sorted(whales_namelist)
for whale_name in whales_namelist:
    print "{} observations: {} pods, {} individuals".format(whale_name, whales_odd[whale_name], whales_ind[whale_name])

File: COMP9041/lab06/all_whales.pl
#!/usr/bin/perl -w

# count all whale observations
# I use 2 hash list to represent the whale_name-pod-ind pair
# key: whale_name

# reads whale observations in the above format until the end-of-input 
@whale_record = ();
while($pod = <STDIN>){
    push @whale_record, $pod;
}

# scan each line
@whales_namelist = ();
%whales_odd = ();
%whales_ind = ();
foreach $pod(@whale_record){
    # split the pod by number and strings
    $pod =~ s/^\s+|\s+$//g; # remove the spaces
    @splitted = split(" +", $pod);
    $splitted_length = @splitted;
    
    $individual_ob = $splitted[0];
    $whale_name = join(' ', @splitted[1..$splitted_length-1]); # join the rest elements as whale name, or [1..$#splitted]
    $whale_name = lc($whale_name); # all lower case
    $whale_name =~ s/s$//; # remove s at last  
    chomp $whale_name; # remove \n
    
    # test: passed
    # print "whale name is $whale_name";
    # print "the observations in this pod is $individual_ob\n";
    
    # check whether this whale has been observed
    if (exists($whales_odd{$whale_name})){
        # print "Has been observed before\n";
        $whales_odd{$whale_name}++;
        $whales_ind{$whale_name} += $individual_ob;
    }
    else{ # new whale
        # print "Has not been observed before\n";
        push @whales_namelist, $whale_name;
        $whales_odd{$whale_name} = 1;
        $whales_ind{$whale_name} = $individual_ob;
    }
}

# output the data in alphabetical order
@whales_namelist = sort(@whales_namelist);
foreach $whale_name(@whales_namelist){
    print "$whale_name observations: $whales_odd{$whale_name} pods, $whales_ind{$whale_name} individuals\n";
}

